### PR TITLE
Properly set the execute_on flag for linear

### LINKED ISSUE #8610: FEProblemBase:: _current_execute_on_flag is not set for EXEC_LINEAR
### Description of the enhancement or error report
This flag is currently set by `FEProblemBase::execute()`, which is not called by `FEProblem::computeResidualType()`. Thus this flag is not set for EXEC_LINEAR.

### Rationale for the enhancement or information for reproducing the error
This is in consistent what this flag is supposed to be doing.

### Identified impact
(i.e. Internal object changes, limited interface changes, public API change, or a list of specific applications impacted)
None. But I will be using this feature soon to check where the material property evaluation is happening.

### PR DESCRIPTION
**Pull request by YaqiWang**
Close #8610.

### CODE REVIEW

**Review by permcody — CHANGES_REQUESTED** [review]
This doesn't look quite right. The Transfers and Multiapps still aren't correct. Can you move the first statement up further?

**Comment by YaqiWang** [comment]
No, because this is mimic the one in `FEProblemBase::execute()`. I am feeling we need to be consistent.

**Comment by permcody** [comment]
Well that one is probably wrong too. If someone were to check that flag in a multiapp it would report something inconsistent. Let's change them both!

**Comment by YaqiWang** [comment]
How about this, set the flag in all relevant subroutines.

**Comment by permcody** [comment]
Hmm, that's a lot more twiddling and will probably get out of date faster. I think I liked your original PR better with the small change I requested. This can go in, but it'll likely break without a slew of new tests depending on that flag being correct.

**Comment by YaqiWang** [comment]
Cannot think of a way of testing this. I actually like the later one. It may not that fragile as you thought.

**Comment by permcody** [comment]
Well you could easily check by creating a whole slew of new objects that look at the flag and incorporate that into the solution. I'm not advocating this right now, but it could be done. Let's go with solution one.

**Comment by YaqiWang** [comment]
How do you want to handle the flag for multiapps?

**Comment by permcody** [comment]
If you simply set the flag earlier than you had it in your first PR, it'll be correct for Multiapps as well (i.e. A MultiApp set to execute on linear will be able to read that flag and see that it's being executed on linear).

**Comment by YaqiWang** [comment]
Changed per you request @permcody. But I still think the second one is more correct.

File: framework/src/base/FEProblemBase.C [hunk 1]
```diff
@@ -3590,6 +3590,8 @@ FEProblemBase::computeResidualType(const NumericVector<Number>& soln, NumericVec
 
   executeControls(EXEC_LINEAR);
 
+  _current_execute_on_flag = EXEC_NONE;
+
   _app.getOutputWarehouse().residualSetup();
 
   _nl->computeResidual(residual, type);
```
Comment (permcody) [inline framework/src/base/FEProblemBase.C]: What about this line? You're setting the flag back before we call compute residual.

Comment (YaqiWang) [inline framework/src/base/FEProblemBase.C]: That is on purpose. Kernels do not have execute_on.

**Comment by permcody** [comment]
If you get around to it a test would be nice. Otherwise this could always get broken in the future.

**Review by permcody — APPROVED** [review]

**Comment by YaqiWang** [comment]
Working on a simple test, I plan to let aux kernel on different execute_ons to expose this flag. Meanwhile I noticed that EXEC_INITIAL is not properly set too.

File: framework/src/base/FEProblemBase.C (hunk 1 above)
Comment (GiudGiud) [reply framework/src/base/FEProblemBase.C:3597]: I dont see why we are unsetting it though ?

This is causing issues for the execution order printing, as the flag is wrong when we print it.

**Pull request merged**